File: src/lib/eudr/satellite-checker.ts
import { createHash } from "node:crypto";
import { geodesicAreaHa } from "./gis-validator";
import { EUDR_CUTOFF_DATE, EUDR_CUTOFF_YEAR, type GeoJsonInput, type RiskLevel, type SatelliteCheckResult, type SupportedGeometry } from "./types";

// ---------------------------------------------------------------- Benchmark pays (Commission UE, art. 29 EUDR)
interface CountryBox {
  iso2: string;
  name: string;
  box: [number, number, number, number]; // lonMin, latMin, lonMax, latMax
  risk: RiskLevel;
}

const COUNTRY_BOXES: CountryBox[] = [
  { iso2: "BY", name: "Biélorussie", box: [23.1, 51.2, 32.8, 56.2], risk: "HIGH" },
  { iso2: "MM", name: "Myanmar", box: [92.1, 9.5, 101.2, 28.6], risk: "HIGH" },
  { iso2: "KP", name: "Corée du Nord", box: [124.1, 37.6, 130.7, 43.0], risk: "HIGH" },
  { iso2: "RU", name: "Russie", box: [27.3, 41.1, 180.0, 81.9], risk: "HIGH" },
  { iso2: "CI", name: "Côte d'Ivoire", box: [-8.6, 4.3, -2.5, 10.8], risk: "STANDARD" },
  { iso2: "GH", name: "Ghana", box: [-3.3, 4.7, 1.2, 11.2], risk: "STANDARD" },
  { iso2: "NG", name: "Nigeria", box: [2.6, 4.2, 14.7, 13.9], risk: "STANDARD" },
  { iso2: "CM", name: "Cameroun", box: [8.4, 1.6, 16.2, 13.1], risk: "STANDARD" },
  { iso2: "CD", name: "RD Congo", box: [12.2, -13.5, 31.3, 5.4], risk: "STANDARD" },
  { iso2: "UG", name: "Ouganda", box: [29.5, -1.5, 35.0, 4.2], risk: "STANDARD" },
  { iso2: "ET", name: "Éthiopie", box: [32.9, 3.4, 48.0, 14.9], risk: "STANDARD" },
  { iso2: "KE", name: "Kenya", box: [33.9, -4.7, 41.9, 5.5], risk: "STANDARD" },
  { iso2: "TZ", name: "Tanzanie", box: [29.3, -11.8, 40.5, -0.9], risk: "STANDARD" },
  { iso2: "LR", name: "Liberia", box: [-11.5, 4.3, -7.4, 8.6], risk: "STANDARD" },
  { iso2: "ID", name: "Indonésie", box: [95.0, -11.0, 141.0, 6.1], risk: "STANDARD" },
  { iso2: "MY", name: "Malaisie", box: [99.6, 0.8, 119.3, 7.4], risk: "STANDARD" },
  { iso2: "VN", name: "Vietnam", box: [102.1, 8.4, 109.5, 23.4], risk: "STANDARD" },
  { iso2: "TH", name: "Thaïlande", box: [97.3, 5.6, 105.7, 20.5], risk: "STANDARD" },
  { iso2: "PG", name: "Papouasie-Nouvelle-Guinée", box: [140.8, -11.7, 156.0, -1.3], risk: "STANDARD" },
  { iso2: "IN", name: "Inde", box: [68.1, 6.7, 97.4, 35.5], risk: "STANDARD" },
  { iso2: "BR", name: "Brésil", box: [-74.0, -33.8, -34.8, 5.3], risk: "STANDARD" },
  { iso2: "CO", name: "Colombie", box: [-79.0, -4.2, -66.9, 12.5], risk: "STANDARD" },
  { iso2: "PE", name: "Pérou", box: [-81.4, -18.4, -68.7, -0.03], risk: "STANDARD" },
  { iso2: "EC", name: "Équateur", box: [-81.1, -5.0, -75.2, 1.7], risk: "STANDARD" },
  { iso2: "BO", name: "Bolivie", box: [-69.6, -22.9, -57.5, -9.7], risk: "STANDARD" },
  { iso2: "PY", name: "Paraguay", box: [-62.7, -27.6, -54.3, -19.3], risk: "STANDARD" },
  { iso2: "AR", name: "Argentine", box: [-73.6, -55.1, -53.6, -21.8], risk: "STANDARD" },
  { iso2: "HN", name: "Honduras", box: [-89.4, 12.9, -83.1, 16.5], risk: "STANDARD" },
  { iso2: "GT", name: "Guatemala", box: [-92.3, 13.7, -88.2, 17.8], risk: "STANDARD" },
  { iso2: "NI", name: "Nicaragua", box: [-87.7, 10.7, -83.1, 15.0], risk: "STANDARD" },
  { iso2: "CR", name: "Costa Rica", box: [-85.9, 8.0, -82.6, 11.2], risk: "STANDARD" },
  { iso2: "MX", name: "Mexique", box: [-118.4, 14.5, -86.7, 32.7], risk: "STANDARD" },
  { iso2: "US", name: "États-Unis", box: [-125.0, 24.5, -66.9, 49.4], risk: "LOW" },
  { iso2: "CA", name: "Canada", box: [-141.0, 41.7, -52.6, 83.1], risk: "LOW" },
  { iso2: "AU", name: "Australie", box: [113.3, -43.6, 153.6, -10.7], risk: "LOW" },
  { iso2: "NZ", name: "Nouvelle-Zélande", box: [166.5, -47.3, 178.6, -34.4], risk: "LOW" },
  { iso2: "JP", name: "Japon", box: [129.4, 31.0, 145.8, 45.5], risk: "LOW" },
  { iso2: "CN", name: "Chine", box: [73.5, 18.2, 134.8, 53.6], risk: "LOW" },
  { iso2: "EU", name: "Union européenne", box: [-10.5, 35.0, 31.6, 71.2], risk: "LOW" },
];

export function resolveCountry(lon: number, lat: number): { iso2: string; name: string; risk: RiskLevel } {
  for (const c of COUNTRY_BOXES) {
    const [lonMin, latMin, lonMax, latMax] = c.box;
    if (lon >= lonMin && lon <= lonMax && lat >= latMin && lat <= latMax) return { iso2: c.iso2, name: c.name, risk: c.risk };
  }
  return { iso2: "XX", name: "Non déterminé", risk: "STANDARD" };
}

// ---------------------------------------------------------------- Hotspots déterministes (mode hors-ligne)
interface LossHotspot {
  label: string;
  box: [number, number, number, number];
  lossYear: number;
  lossFraction: number;
  treeCover2000: number;
}

const LOSS_HOTSPOTS: LossHotspot[] = [
  { label: "Arc de déforestation — Pará (BR)", box: [-56.0, -10.0, -48.0, -2.0], lossYear: 2022, lossFraction: 0.42, treeCover2000: 88 },
  { label: "Riau — Sumatra (ID)", box: [100.0, -1.0, 104.0, 2.0], lossYear: 2021, lossFraction: 0.35, treeCover2000: 81 },
  { label: "Kalimantan central (ID)", box: [110.0, -3.0, 117.0, 2.0], lossYear: 2024, lossFraction: 0.27, treeCover2000: 84 },
  { label: "Sud-ouest ivoirien — Taï (CI)", box: [-8.0, 5.0, -6.0, 7.0], lossYear: 2023, lossFraction: 0.18, treeCover2000: 76 },
  { label: "Mato Grosso (BR) — perte pré-cutoff", box: [-60.0, -16.0, -52.0, -10.01], lossYear: 2019, lossFraction: 0.3, treeCover2000: 79 },
  { label: "Bassin du Congo — Tshopo (CD)", box: [24.0, -1.0, 27.0, 2.0], lossYear: 2019, lossFraction: 0.12, treeCover2000: 90 },
];

function stableFraction(seed: string): number {
  const digest = createHash("sha256").update(seed).digest("hex");
  return parseInt(digest.slice(0, 8), 16) / 0xffffffff;
}

function extractSimulatedLossYear(input: GeoJsonInput): number | null {
  const props = input.properties as Record<string, unknown> | undefined;
  if (props && props.simulated_loss_year !== undefined && props.simulated_loss_year !== null) {
    const y = Number(props.simulated_loss_year);
    return Number.isFinite(y) ? Math.trunc(y) : null;
  }
  if (input.type === "FeatureCollection" && Array.isArray(input.features)) {
    for (const f of input.features as GeoJsonInput[]) {
      const y = extractSimulatedLossYear(f);
      if (y !== null) return y;
    }
  }
  return null;
}

function centroidOf(geometry: SupportedGeometry): [number, number] {
  const pts: number[][] = [];
  const walk = (c: unknown): void => {
    if (!Array.isArray(c)) return;
    if (typeof c[0] === "number") {
      pts.push(c as number[]);
      return;
    }
    c.forEach(walk);
  };
  walk(geometry.coordinates);
  const sum = pts.reduce((acc, p) => [acc[0] + p[0], acc[1] + p[1]], [0, 0]);
  return [sum[0] / pts.length, sum[1] / pts.length];
}

function bufferPoint(geometry: SupportedGeometry): SupportedGeometry {
  if (geometry.type !== "Point" && geometry.type !== "MultiPoint") return geometry;
  const [lon, lat] = centroidOf(geometry);
  const r = 0.0005; // ≈ 55 m
  const ring: number[][] = [];
  for (let i = 0; i <= 32; i++) {
    const a = (i / 32) * 2 * Math.PI;
    ring.push([lon + r * Math.cos(a), lat + r * Math.sin(a)]);
  }
  return { type: "Polygon", coordinates: [ring] };
}

// ---------------------------------------------------------------- Moteur déterministe
export function deterministicCheck(geometry: SupportedGeometry, harvestDate: string, forcedLossYear: number | null): SatelliteCheckResult {
  const [lon, lat] = centroidOf(geometry);
  const country = resolveCountry(lon, lat);
  const jitter = stableFraction(`${lon.toFixed(4)}:${lat.toFixed(4)}`);

  let lossYear: number | null = forcedLossYear;
  let lossFraction = 0;
  let treeCover = Number((35 + jitter * 40).toFixed(1));
  let hotspotLabel: string | null = null;

  if (lossYear === null) {
    for (const spot of LOSS_HOTSPOTS) {
      const [lonMin, latMin, lonMax, latMax] = spot.box;
      if (lon >= lonMin && lon <= lonMax && lat >= latMin && lat <= latMax) {
        lossYear = spot.lossYear;
        lossFraction = spot.lossFraction;
        treeCover = spot.treeCover2000;
        hotspotLabel = spot.label;
        break;
      }
    }
  } else {
    lossFraction = 0.25;
    hotspotLabel = "Année de perte simulée (mode démonstration)";
  }

  const areaHa = geodesicAreaHa(geometry);
  const lossArea = areaHa > 0 ? Number((areaHa * lossFraction).toFixed(4)) : lossYear ? 0.05 : 0;
  const postCutoff = lossYear !== null && lossYear > EUDR_CUTOFF_YEAR;

  let confidence: number;
  let riskLevel: RiskLevel;
  let details: string;

  if (postCutoff) {
    confidence = Number((0.86 + jitter * 0.12).toFixed(3));
    riskLevel = "HIGH";
    details = `Perte de couvert forestier détectée en ${lossYear} (${lossArea} ha, ${Math.round(lossFraction * 100)} % de la parcelle) — postérieure au 31/12/2020. Zone : ${hotspotLabel}.`;
  } else if (lossYear !== null) {
    confidence = Number((0.88 + jitter * 0.1).toFixed(3));
    riskLevel = country.risk === "LOW" ? "STANDARD" : country.risk;
    details = `Perte historique en ${lossYear} (antérieure à la date butoir) : conforme, mais vigilance renforcée. Zone : ${hotspotLabel}.`;
  } else {
    confidence = Number((0.9 + jitter * 0.09).toFixed(3));
    riskLevel = country.risk;
    details = `Aucune alerte Hansen Tree Cover Loss post-2020 sur l'emprise. Pays : ${country.name} (${country.iso2}), benchmark UE : ${country.risk}.`;
  }

  if (harvestDate <= EUDR_CUTOFF_DATE) details += " Récolte antérieure à la date butoir : hors champ temporel EUDR.";

  return {
    compliant: !postCutoff,
    loss_year: lossYear,
    confidence_score: Math.min(confidence, 1),
    risk_level: riskLevel,
    country_code: country.iso2,
    country_name: country.name,
    country_risk: country.risk,
    source: "deterministic-mock (Hansen/GFW rules engine)",
    loss_area_ha: lossArea,
    tree_cover_2000_pct: treeCover,
    details,
  };
}

// ---------------------------------------------------------------- Moteur live Global Forest Watch
interface GfwRow {
  umd_tree_cover_loss__year?: number | string | null;
  area__ha?: number | string | null;
}

async function gfwQuery(geometry: SupportedGeometry): Promise<GfwRow[]> {
  const base = (process.env.GFW_API_URL ?? "https://data-api.globalforestwatch.org").replace(/\/$/, "");
  const dataset = process.env.GFW_DATASET ?? "umd_tree_cover_loss";
  const sql =
    "SELECT umd_tree_cover_loss__year, SUM(area__ha) AS area__ha FROM results " +
    "WHERE umd_tree_cover_density_2000__threshold = 30 GROUP BY umd_tree_cover_loss__year ORDER BY umd_tree_cover_loss__year";
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), Number(process.env.GFW_TIMEOUT_SECONDS ?? "15") * 1000);
  try {
    const res = await fetch(`${base}/dataset/${dataset}/latest/query`, {
      method: "POST",
      headers: { "x-api-key": process.env.GFW_API_KEY ?? "", "Content-Type": "application/json" },
      body: JSON.stringify({ sql, geometry: bufferPoint(geometry) }),
      signal: controller.signal,
      cache: "no-store",
    });
    if (!res.ok) throw new Error(`GFW HTTP ${res.status}`);
    const payload = (await res.json()) as { data?: GfwRow[] };
    return payload.data ?? [];
  } finally {
    clearTimeout(timeout);
  }
}

async function liveCheck(geometry: SupportedGeometry): Promise<SatelliteCheckResult> {
  const rows = await gfwQuery(geometry);
  const [lon, lat] = centroidOf(geometry);
  const country = resolveCountry(lon, lat);
  const areaHa = geodesicAreaHa(geometry) || 1;
  const parsed = rows
    .map((r) => ({ year: Number(r.umd_tree_cover_loss__year ?? 0), area: Number(r.area__ha ?? 0) }))
    .filter((r) => r.year > 0);
  const post = parsed.filter((r) => r.year > EUDR_CUTOFF_YEAR);
  const lossArea = Number(post.reduce((acc, r) => acc + r.area, 0).toFixed(4));
  const significant = lossArea >= Math.max(0.01, areaHa * 0.005);

  if (significant) {
    const lossYear = Math.min(...post.map((r) => r.year));
    const fraction = Math.min(lossArea / areaHa, 1);
    return {
      compliant: false,
      loss_year: lossYear,
      confidence_score: Number(Math.min(0.99, 0.8 + fraction * 0.5).toFixed(3)),
      risk_level: "HIGH",
      country_code: country.iso2,
      country_name: country.name,
      country_risk: country.risk,
      source: "gfw-live (umd_tree_cover_loss)",
      loss_area_ha: lossArea,
      tree_cover_2000_pct: null,
      details: `GFW : ${lossArea} ha de perte détectés à partir de ${lossYear} (post-2020) sur ${areaHa.toFixed(2)} ha.`,
    };
  }
  const historic = parsed.length ? Math.max(...parsed.map((r) => r.year)) : null;
  return {
    compliant: true,
    loss_year: historic,
    confidence_score: 0.95,
    risk_level: country.risk,
    country_code: country.iso2,
    country_name: country.name,
    country_risk: country.risk,
    source: "gfw-live (umd_tree_cover_loss)",
    loss_area_ha: lossArea,
    tree_cover_2000_pct: null,
    details: `GFW : aucune perte significative post-2020 (${lossArea} ha). Pays : ${country.name}, benchmark UE : ${country.risk}.`,
  };
}

export function gfwLiveAvailable(): boolean {
  const enabled = (process.env.GFW_LIVE_ENABLED ?? "true").toLowerCase() !== "false";
  return enabled && Boolean(process.env.GFW_API_KEY);
}

/**
 * Évalue le risque de déforestation post-2020 sur une géométrie normalisée.
 * @param originalInput GeoJSON d'origine (pour lire d'éventuelles propriétés de démonstration).
 */
export async function checkDeforestationRisk(
  geometry: SupportedGeometry,
  harvestDate: string,
  originalInput: GeoJsonInput,
): Promise<SatelliteCheckResult> {
  const forced = extractSimulatedLossYear(originalInput);
  if (gfwLiveAvailable() && forced === null) {
    try {
      return await liveCheck(geometry);
    } catch (err) {
      console.warn("[GFW] live indisponible — repli déterministe :", (err as Error).message);
    }
  }
  return deterministicCheck(geometry, harvestDate, forced);
}

File: backend/app/services/satellite_checker.py
"""Détection de déforestation post-2020 (Hansen / UMD Tree Cover Loss via Global Forest Watch).

Deux moteurs :
  1. **live**  : GFW Data API (`/dataset/umd_tree_cover_loss/latest/query`) si `GFW_API_KEY` est configurée.
  2. **mock déterministe** : règles de scoring reproductibles (hotspots documentés + benchmark pays UE)
     utilisées hors-ligne, en tests et en repli automatique si l'API est indisponible.

Règle EUDR (art. 2(13) & art. 3) : toute perte de couvert forestier datée APRÈS le 31/12/2020
sur l'emprise de la parcelle => `compliant = False`.
"""
from __future__ import annotations

import hashlib
import json
import logging
from dataclasses import dataclass
from datetime import date
from typing import Any, Dict, List, Optional, Tuple

import requests
from shapely.geometry import shape
from shapely.geometry.base import BaseGeometry

from app.core.config import settings

logger = logging.getLogger(__name__)

RiskLevel = str  # "LOW" | "STANDARD" | "HIGH"


# --------------------------------------------------------------------------- Benchmark pays
@dataclass(frozen=True)
class CountryBox:
    iso2: str
    name: str
    lon_min: float
    lat_min: float
    lon_max: float
    lat_max: float
    risk: RiskLevel


# Emprises approximatives (ordre = priorité). Benchmark conforme à la classification
# publiée par la Commission (règlement d'exécution 2025, art. 29 EUDR) :
#   HIGH : Biélorussie, Myanmar, Corée du Nord, Russie — LOW : UE et la plupart des pays
#   à faible risque — STANDARD : le reste (grands pays producteurs tropicaux).
COUNTRY_BOXES: List[CountryBox] = [
    CountryBox("BY", "Biélorussie", 23.1, 51.2, 32.8, 56.2, "HIGH"),
    CountryBox("MM", "Myanmar", 92.1, 9.5, 101.2, 28.6, "HIGH"),
    CountryBox("KP", "Corée du Nord", 124.1, 37.6, 130.7, 43.0, "HIGH"),
    CountryBox("RU", "Russie", 27.3, 41.1, 180.0, 81.9, "HIGH"),
    CountryBox("CI", "Côte d'Ivoire", -8.6, 4.3, -2.5, 10.8, "STANDARD"),
    CountryBox("GH", "Ghana", -3.3, 4.7, 1.2, 11.2, "STANDARD"),
    CountryBox("NG", "Nigeria", 2.6, 4.2, 14.7, 13.9, "STANDARD"),
    CountryBox("CM", "Cameroun", 8.4, 1.6, 16.2, 13.1, "STANDARD"),
    CountryBox("CD", "RD Congo", 12.2, -13.5, 31.3, 5.4, "STANDARD"),
    CountryBox("UG", "Ouganda", 29.5, -1.5, 35.0, 4.2, "STANDARD"),
    CountryBox("ET", "Éthiopie", 32.9, 3.4, 48.0, 14.9, "STANDARD"),
    CountryBox("KE", "Kenya", 33.9, -4.7, 41.9, 5.5, "STANDARD"),
    CountryBox("TZ", "Tanzanie", 29.3, -11.8, 40.5, -0.9, "STANDARD"),
    CountryBox("LR", "Liberia", -11.5, 4.3, -7.4, 8.6, "STANDARD"),
    CountryBox("ID", "Indonésie", 95.0, -11.0, 141.0, 6.1, "STANDARD"),
    CountryBox("MY", "Malaisie", 99.6, 0.8, 119.3, 7.4, "STANDARD"),
    CountryBox("VN", "Vietnam", 102.1, 8.4, 109.5, 23.4, "STANDARD"),
    CountryBox("TH", "Thaïlande", 97.3, 5.6, 105.7, 20.5, "STANDARD"),
    CountryBox("PG", "Papouasie-Nouvelle-Guinée", 140.8, -11.7, 156.0, -1.3, "STANDARD"),
    CountryBox("IN", "Inde", 68.1, 6.7, 97.4, 35.5, "STANDARD"),
    CountryBox("BR", "Brésil", -74.0, -33.8, -34.8, 5.3, "STANDARD"),
    CountryBox("CO", "Colombie", -79.0, -4.2, -66.9, 12.5, "STANDARD"),
    CountryBox("PE", "Pérou", -81.4, -18.4, -68.7, -0.03, "STANDARD"),
    CountryBox("EC", "Équateur", -81.1, -5.0, -75.2, 1.7, "STANDARD"),
    CountryBox("BO", "Bolivie", -69.6, -22.9, -57.5, -9.7, "STANDARD"),
    CountryBox("PY", "Paraguay", -62.7, -27.6, -54.3, -19.3, "STANDARD"),
    CountryBox("AR", "Argentine", -73.6, -55.1, -53.6, -21.8, "STANDARD"),
    CountryBox("HN", "Honduras", -89.4, 12.9, -83.1, 16.5, "STANDARD"),
    CountryBox("GT", "Guatemala", -92.3, 13.7, -88.2, 17.8, "STANDARD"),
    CountryBox("NI", "Nicaragua", -87.7, 10.7, -83.1, 15.0, "STANDARD"),
    CountryBox("CR", "Costa Rica", -85.9, 8.0, -82.6, 11.2, "STANDARD"),
    CountryBox("MX", "Mexique", -118.4, 14.5, -86.7, 32.7, "STANDARD"),
    CountryBox("US", "États-Unis", -125.0, 24.5, -66.9, 49.4, "LOW"),
    CountryBox("CA", "Canada", -141.0, 41.7, -52.6, 83.1, "LOW"),
    CountryBox("AU", "Australie", 113.3, -43.6, 153.6, -10.7, "LOW"),
    CountryBox("NZ", "Nouvelle-Zélande", 166.5, -47.3, 178.6, -34.4, "LOW"),
    CountryBox("JP", "Japon", 129.4, 31.0, 145.8, 45.5, "LOW"),
    CountryBox("CN", "Chine", 73.5, 18.2, 134.8, 53.6, "LOW"),
    CountryBox("EU", "Union européenne", -10.5, 35.0, 31.6, 71.2, "LOW"),
]


def resolve_country(lon: float, lat: float) -> Tuple[str, str, RiskLevel]:
    """Retourne (iso2, nom, niveau de risque benchmark) pour un centroïde."""
    for box in COUNTRY_BOXES:
        if box.lon_min <= lon <= box.lon_max and box.lat_min <= lat <= box.lat_max:
            return box.iso2, box.name, box.risk
    return "XX", "Non déterminé", "STANDARD"


# --------------------------------------------------------------------------- Hotspots déterministes
@dataclass(frozen=True)
class LossHotspot:
    label: str
    lon_min: float
    lat_min: float
    lon_max: float
    lat_max: float
    loss_year: int
    loss_fraction: float  # fraction de la parcelle touchée
    tree_cover_2000_pct: float


# Zones de pression documentées (arc de déforestation, Riau, Kalimantan, sud-ouest ivoirien...).
# Utilisées UNIQUEMENT par le moteur déterministe (hors-ligne / tests).
LOSS_HOTSPOTS: List[LossHotspot] = [
    LossHotspot("Arc de déforestation — Pará (BR)", -56.0, -10.0, -48.0, -2.0, 2022, 0.42, 88.0),
    LossHotspot("Riau — Sumatra (ID)", 100.0, -1.0, 104.0, 2.0, 2021, 0.35, 81.0),
    LossHotspot("Kalimantan central (ID)", 110.0, -3.0, 117.0, 2.0, 2024, 0.27, 84.0),
    LossHotspot("Sud-ouest ivoirien — Taï (CI)", -8.0, 5.0, -6.0, 7.0, 2023, 0.18, 76.0),
    LossHotspot("Mato Grosso (BR) — perte pré-cutoff", -60.0, -16.0, -52.0, -10.01, 2019, 0.30, 79.0),
    LossHotspot("Bassin du Congo — Tshopo (CD)", 24.0, -1.0, 27.0, 2.0, 2019, 0.12, 90.0),
]


def _stable_fraction(seed: str) -> float:
    """Pseudo-aléa déterministe dans [0, 1) dérivé d'un hash SHA-256."""
    digest = hashlib.sha256(seed.encode("utf-8")).hexdigest()
    return int(digest[:8], 16) / 0xFFFFFFFF


def _extract_simulated_loss_year(geometry: Dict[str, Any]) -> Optional[int]:
    """Permet aux jeux de démonstration d'imposer une année de perte via `properties.simulated_loss_year`."""
    props = geometry.get("properties") if isinstance(geometry, dict) else None
    if isinstance(props, dict) and props.get("simulated_loss_year") is not None:
        try:
            return int(props["simulated_loss_year"])
        except (TypeError, ValueError):
            return None
    if geometry.get("type") == "FeatureCollection":
        for feature in geometry.get("features") or []:
            year = _extract_simulated_loss_year(feature)
            if year is not None:
                return year
    return None


def _to_shapely(geometry: Dict[str, Any]) -> BaseGeometry:
    gtype = geometry.get("type")
    if gtype == "Feature":
        return shape(geometry["geometry"])
    if gtype == "FeatureCollection":
        from shapely.ops import unary_union

        return unary_union([shape(f["geometry"]) for f in geometry["features"] if f.get("geometry")])
    return shape(geometry)


def _analysis_polygon(geom: BaseGeometry) -> Dict[str, Any]:
    """Pour un point, on analyse un disque ~55 m (0.0005°) ; sinon la géométrie elle-même."""
    if geom.geom_type in {"Point", "MultiPoint"}:
        geom = geom.buffer(0.0005)
    from shapely.geometry import mapping

    return mapping(geom)


# --------------------------------------------------------------------------- Moteur déterministe
def deterministic_check(geom: BaseGeometry, harvest_date: date, forced_loss_year: Optional[int] = None) -> Dict[str, Any]:
    centroid = geom.centroid
    lon, lat = centroid.x, centroid.y
    iso2, country_name, country_risk = resolve_country(lon, lat)
    seed = f"{round(lon, 4)}:{round(lat, 4)}"
    jitter = _stable_fraction(seed)

    loss_year: Optional[int] = forced_loss_year
    loss_fraction = 0.0
    tree_cover = round(35.0 + jitter * 40.0, 1)
    hotspot_label: Optional[str] = None

    if loss_year is None:
        for spot in LOSS_HOTSPOTS:
            if spot.lon_min <= lon <= spot.lon_max and spot.lat_min <= lat <= spot.lat_max:
                loss_year = spot.loss_year
                loss_fraction = spot.loss_fraction
                tree_cover = spot.tree_cover_2000_pct
                hotspot_label = spot.label
                break
    elif forced_loss_year is not None:
        loss_fraction = 0.25
        hotspot_label = "Année de perte simulée (mode démonstration)"

    from app.services.gis_validator import geodesic_area_ha

    area_ha = geodesic_area_ha(geom)
    loss_area = round(area_ha * loss_fraction, 4) if area_ha > 0 else (0.05 if loss_year else 0.0)

    post_cutoff = loss_year is not None and loss_year > settings.eudr_cutoff_date.year
    compliant = not post_cutoff

    if post_cutoff:
        confidence = round(0.86 + jitter * 0.12, 3)
        risk_level: RiskLevel = "HIGH"
        details = (
            f"Perte de couvert forestier détectée en {loss_year} ({loss_area} ha, {loss_fraction:.0%} de la parcelle)"
            f" — postérieure au 31/12/2020. Zone : {hotspot_label}."
        )
    elif loss_year is not None:
        confidence = round(0.88 + jitter * 0.10, 3)
        risk_level = country_risk if country_risk != "LOW" else "STANDARD"
        details = (
            f"Perte historique en {loss_year} (antérieure à la date butoir) : conforme, mais vigilance renforcée. "
            f"Zone : {hotspot_label}."
        )
    else:
        confidence = round(0.90 + jitter * 0.09, 3)
        risk_level = country_risk
        details = (
            f"Aucune alerte Hansen Tree Cover Loss post-2020 sur l'emprise. Pays : {country_name} ({iso2}), "
            f"benchmark UE : {country_risk}."
        )

    if harvest_date <= settings.eudr_cutoff_date:
        details += " Récolte antérieure à la date butoir : hors champ temporel EUDR."

    return {
        "compliant": compliant,
        "loss_year": loss_year,
        "confidence_score": min(confidence, 1.0),
        "risk_level": risk_level,
        "country_code": iso2,
        "country_risk": country_risk,
        "source": "deterministic-mock (Hansen/GFW rules engine)",
        "loss_area_ha": loss_area,
        "tree_cover_2000_pct": tree_cover,
        "details": details,
    }


# --------------------------------------------------------------------------- Moteur live GFW
def _gfw_query(analysis_geometry: Dict[str, Any]) -> List[Dict[str, Any]]:
    """Interroge GFW Data API : pertes annuelles (ha) sur l'emprise, densité 2000 >= 30 %."""
    url = f"{settings.gfw_api_url.rstrip('/')}/dataset/{settings.gfw_dataset}/latest/query"
    sql = (
        "SELECT umd_tree_cover_loss__year, SUM(area__ha) AS area__ha "
        "FROM results WHERE umd_tree_cover_density_2000__threshold = 30 "
        "GROUP BY umd_tree_cover_loss__year ORDER BY umd_tree_cover_loss__year"
    )
    response = requests.post(
        url,
        headers={"x-api-key": settings.gfw_api_key, "Content-Type": "application/json"},
        data=json.dumps({"sql": sql, "geometry": analysis_geometry}),
        timeout=settings.gfw_timeout_seconds,
    )
    response.raise_for_status()
    payload = response.json()
    return list(payload.get("data") or [])


def live_check(geom: BaseGeometry, harvest_date: date) -> Dict[str, Any]:
    rows = _gfw_query(_analysis_polygon(geom))
    centroid = geom.centroid
    iso2, country_name, country_risk = resolve_country(centroid.x, centroid.y)

    from app.services.gis_validator import geodesic_area_ha

    area_ha = geodesic_area_ha(geom) or 1.0
    post_cutoff_rows = [
        r for r in rows if int(r.get("umd_tree_cover_loss__year") or 0) > settings.eudr_cutoff_date.year
    ]
    all_rows = [r for r in rows if r.get("umd_tree_cover_loss__year")]

    loss_area = round(sum(float(r.get("area__ha") or 0.0) for r in post_cutoff_rows), 4)
    # Seuil de bruit : 0,5 % de la parcelle ou 0,01 ha (~1 pixel Landsat = 0,09 ha, on reste strict)
    significant = loss_area >= max(0.01, area_ha * 0.005)

    if significant:
        loss_year = min(int(r["umd_tree_cover_loss__year"]) for r in post_cutoff_rows)
        fraction = min(loss_area / area_ha, 1.0)
        confidence = round(min(0.99, 0.80 + fraction * 0.5), 3)
        return {
            "compliant": False,
            "loss_year": loss_year,
            "confidence_score": confidence,
            "risk_level": "HIGH",
            "country_code": iso2,
            "country_risk": country_risk,
            "source": "gfw-live (umd_tree_cover_loss)",
            "loss_area_ha": loss_area,
            "tree_cover_2000_pct": None,
            "details": f"GFW : {loss_area} ha de perte détectés à partir de {loss_year} (post-2020) sur {area_ha:.2f} ha.",
        }

    historic_year = max((int(r["umd_tree_cover_loss__year"]) for r in all_rows), default=None)
    return {
        "compliant": True,
        "loss_year": historic_year,
        "confidence_score": 0.95,
        "risk_level": country_risk,
        "country_code": iso2,
        "country_risk": country_risk,
        "source": "gfw-live (umd_tree_cover_loss)",
        "loss_area_ha": loss_area,
        "tree_cover_2000_pct": None,
        "details": (
            f"GFW : aucune perte significative post-2020 ({loss_area} ha). Pays : {country_name}, benchmark UE : {country_risk}."
        ),
    }


# --------------------------------------------------------------------------- API publique
def check_deforestation_risk(geometry: Dict[str, Any], harvest_date: str) -> Dict[str, Any]:
    """Évalue le risque de déforestation post-2020 sur une géométrie GeoJSON.

    :param geometry: Geometry / Feature / FeatureCollection GeoJSON (WGS84).
    :param harvest_date: date de récolte ISO 8601 (YYYY-MM-DD).
    :return: dict conforme à `SatelliteCheckResult` :
             {compliant, loss_year, confidence_score, risk_level, ...}
    """
    parsed_date = date.fromisoformat(harvest_date)
    geom = _to_shapely(geometry)
    forced_year = _extract_simulated_loss_year(geometry)

    if settings.gfw_live_available and forced_year is None:
        try:
            return live_check(geom, parsed_date)
        except (requests.RequestException, ValueError, KeyError) as exc:
            logger.warning("GFW live indisponible (%s) — repli sur le moteur déterministe", exc)

    return deterministic_check(geom, parsed_date, forced_year)
